codegen/cmd/injection-gen/generators/reconciler_reconciler.go

```diff
@@ -367,18 +367,41 @@ func (r *reconcilerImpl) updateStatus(existing *{{.type|raw}}, desired *{{.type|
 			}
 		}
 
+		fullUpdate := false
+		statusUpdate := false
+
+		if !reflect.DeepEqual(existing.GetAnnotations(),desired.GetAnnotations()) {
+			existing.ObjectMeta.Annotations = desired.ObjectMeta.Annotations
+			fullUpdate = true
+		}
+
+		if !reflect.DeepEqual(existing.GetLabels(),desired.GetLabels()) {
+			existing.ObjectMeta.Labels = desired.ObjectMeta.Labels
+			fullUpdate = true
+		}
+
+		if !reflect.DeepEqual(existing.Status, desired.Status) {
+			existing.Status = desired.Status
+			statusUpdate = true
+		}
+
 		// If there's nothing to update, just return.
-		if reflect.DeepEqual(existing.Status, desired.Status) {
+		if !fullUpdate && !statusUpdate {
 			return nil
 		}
 
-		existing.Status = desired.Status
-
 		{{if .nonNamespaced}}
 		updater := r.Client.{{.group}}{{.version}}().{{.type|apiGroup}}()
 		{{else}}
 		updater := r.Client.{{.group}}{{.version}}().{{.type|apiGroup}}(existing.Namespace)
 		{{end}}
+
+		if fullUpdate {
+			_, err = updater.Update(existing)
```

Comment:
this makes me nervous to be honest. It is not clear that it is ok for a controller to add labels and annotations. You should do that in a mutating webhook.

Reply:
@mattmoor wdyt?

Reply:
I wonder if instead of `Update` we should use `Patch`. That would make me less nervous.

Reply:
A few things:
1. With the `/status` sub-resource, I don't believe `Update` will actually update the status, so as written this is probably breaking.
2. I agree with Scotty's sentiment above, that in general this is an anti-pattern and we've been trying to use `// +genreconciler` to codify best practices.  In the past we (Serving) have only used this as an escape hatch to "touch" the resource, but expanded generally I can see a number of problems with this.

For example, some parent resources are particular about the labels and annotations on child resources, so by doing this, we make it possible (maybe even likely) that a child controller will fight the parent controller.  Combined with the bug from `1.` this could result in no status ever really getting published because the `UpdateStatus` calls are always elided in favor of `Update`s 😬 

If you had to do this, I'd recommend using `Patch` from the `ReconcileKind` as @lionelvillard suggests, but I'd be very careful about spinning resource mutations (these can be incredibly hard to detect, especially when the mutations are outside of `spec:` where `metadata.generation` soaring is a tell-tale sign).

Reply:
I'm also not convinced the injected reconciler should do this on the user's behalf. I think it's the role of the user to decide which extraneous values are OK to have inside `current`, and which are not.
This is a slippery slope.